File: hooks/cursor/hooks/enkrypt_guardrails.py
#!/usr/bin/env python
"""
Enkrypt AI Guardrails - Shared Module

This module provides the core Enkrypt API integration used by all hooks:
- beforeSubmitPrompt (input guardrails)
- beforeMCPExecution (input guardrails)
- afterMCPExecution (output guardrails)
- afterAgentResponse (output guardrails)
- stop

Configuration is loaded from guardrails_config.json
"""
# Suppress requests library warnings (urllib3/chardet version mismatches)
import warnings
warnings.filterwarnings("ignore", message="urllib3.*or chardet.*doesn't match")
warnings.filterwarnings("ignore", category=DeprecationWarning)

import os
import json
import datetime
from pathlib import Path

# ============================================================================
# CONFIGURATION LOADING
# ============================================================================

CONFIG_FILE = Path(__file__).parent / "guardrails_config.json"
LOG_DIR = Path(os.environ.get("CURSOR_HOOKS_LOG_DIR", Path.home() / "cursor" / "hooks_logs"))
LOG_DIR.mkdir(parents=True, exist_ok=True)


def load_config() -> dict:
    """Load configuration from guardrails_config.json"""
    try:
        with open(CONFIG_FILE, "r", encoding="utf-8") as f:
            return json.load(f)
    except (FileNotFoundError, json.JSONDecodeError) as e:
        log_event("config_error", {"error": str(e), "config_file": str(CONFIG_FILE)})
        return {}


# Load config on module import
CONFIG = load_config()

# Extract configuration values (strip whitespace to avoid URL issues)
ENKRYPT_API_URL = os.environ.get(
    "ENKRYPT_API_URL",
    CONFIG.get("enkrypt_api", {}).get("url", "https://api.enkryptai.com/guardrails/policy/detect")
).strip()
ENKRYPT_API_KEY = os.environ.get(
    "ENKRYPT_API_KEY",
    CONFIG.get("enkrypt_api", {}).get("api_key", "")
).strip()

SENSITIVE_MCP_TOOLS = CONFIG.get("sensitive_mcp_tools", [])

# Hook-specific policies
HOOK_POLICIES = {
    "beforeSubmitPrompt": CONFIG.get("beforeSubmitPrompt", {}),
    "beforeMCPExecution": CONFIG.get("beforeMCPExecution", {}),
    "afterMCPExecution": CONFIG.get("afterMCPExecution", {}),
    "afterAgentResponse": CONFIG.get("afterAgentResponse", {}),
}


# ============================================================================
# POLICY FUNCTIONS
# ============================================================================

def get_hook_policy(hook_name: str) -> dict:
    """Get policy config for a specific hook."""
    return HOOK_POLICIES.get(hook_name, {})


def is_hook_enabled(hook_name: str) -> bool:
    """Check if guardrails are enabled for a specific hook."""
    return get_hook_policy(hook_name).get("enabled", False)


def get_hook_block_list(hook_name: str) -> list:
    """Get list of detectors that should block for a specific hook."""
    return get_hook_policy(hook_name).get("block", [])


def get_hook_policy_name(hook_name: str) -> str:
    """Get policy name for a specific hook."""
    return get_hook_policy(hook_name).get("policy_name", f"Default {hook_name} Policy")


# ============================================================================
# LOGGING FUNCTIONS
# ============================================================================

def get_timestamp():
    """Get ISO format timestamp."""
    return datetime.datetime.now().isoformat()


def log_event(hook_name: str, data: dict, result: dict = None):
    """Log a hook event to a JSON Lines file."""
    log_file = LOG_DIR / f"{hook_name}.jsonl"

    entry = {
        "timestamp": get_timestamp(),
        "hook": hook_name,
        "input": data,
    }
    if result is not None:
        entry["output"] = result

    with open(log_file, "a", encoding="utf-8") as f:
        f.write(json.dumps(entry) + "\n")


def log_to_combined(hook_name: str, data: dict, result: dict = None):
    """Log to a combined audit log file."""
    log_file = LOG_DIR / "combined_audit.jsonl"

    entry = {
        "timestamp": get_timestamp(),
        "hook": hook_name,
        "conversation_id": data.get("conversation_id"),
        "generation_id": data.get("generation_id"),
        "model": data.get("model"),
        "user_email": data.get("user_email"),
        "data": data,
    }
    if result is not None:
        entry["result"] = result

    with open(log_file, "a", encoding="utf-8") as f:
        f.write(json.dumps(entry) + "\n")


def log_security_alert(alert_type: str, details: dict, data: dict):
    """Log a security alert."""
    alert_file = LOG_DIR / "security_alerts.jsonl"
    alert = {
        "timestamp": get_timestamp(),
        "type": alert_type,
        **details,
        "conversation_id": data.get("conversation_id"),
        "user_email": data.get("user_email"),
    }
    with open(alert_file, "a", encoding="utf-8") as f:
        f.write(json.dumps(alert) + "\n")


# ============================================================================
# ENKRYPT API RESPONSE PARSING
# ============================================================================

def parse_enkrypt_response(result: dict, block_list: list) -> list:
    """
    Parse Enkrypt API response and extract violations.

    API Response Format:
    {
      "summary": {
        "on_topic": 0,
        "nsfw": 1,
        "toxicity": ["toxicity", "insult"],
        "pii": 1,
        "injection_attack": 1,
        "keyword_detected": 1,
        "policy_violation": 1,
        "bias": 1,
        "sponge_attack": 1
      },
      "details": {
        "topic_detector": {...},
        "nsfw": {...},
        "toxicity": {...},
        "pii": {...},
        "injection_attack": {...},
        "keyword_detector": {...},
        "policy_violation": {...},
        "bias": {...},
        "sponge_attack": {...}
      }
    }
    """
    violations = []
    summary = result.get("summary", {})
    details = result.get("details", {})

    # Map summary keys to detector names in block list
    detector_mapping = {
        "nsfw": "nsfw",
        "toxicity": "toxicity",
        "pii": "pii",
        "injection_attack": "injection_attack",
        "keyword_detected": "keyword_detector",
        "policy_violation": "policy_violation",
        "bias": "bias",
        "sponge_attack": "sponge_attack",
        "on_topic": "topic_detector",
    }

    for summary_key, detector_name in detector_mapping.items():
        summary_value = summary.get(summary_key)

        # Check if this detector triggered
        is_detected = False
        if isinstance(summary_value, int) and summary_value == 1:
            is_detected = True
        elif isinstance(summary_value, list) and len(summary_value) > 0:
            is_detected = True
        elif isinstance(summary_value, bool) and summary_value:
            is_detected = True

        # Special case: on_topic=0 means OFF topic (potential issue), on_topic=1 means ON topic (ok)
        if summary_key == "on_topic":
            is_detected = summary_value == 0 if isinstance(summary_value, int) else False

        if is_detected and detector_name in block_list:
            violation_info = {
                "detector": detector_name,
                "detected": True,
                "blocked": True,
            }

            # Add details based on detector type
            detail = details.get(detector_name, {}) or details.get(summary_key, {})

            if detector_name == "pii":
                pii_detail = detail.get("pii", {})
                if pii_detail:
                    violation_info["entities"] = list(pii_detail.keys())
                    violation_info["pii_found"] = pii_detail

            elif detector_name == "toxicity":
                if isinstance(summary_value, list):
                    violation_info["toxicity_types"] = summary_value
                # Get toxicity score
                if "toxicity" in detail:
                    violation_info["score"] = detail.get("toxicity")

            elif detector_name == "keyword_detector":
                violation_info["matched_keywords"] = detail.get("detected_keywords", [])
                violation_info["keyword_counts"] = detail.get("detected_counts", {})

            elif detector_name == "policy_violation":
                violation_info["violating_policy"] = detail.get("violating_policy", "")
                violation_info["explanation"] = detail.get("explanation", "")

            elif detector_name == "injection_attack":
                violation_info["attack_score"] = detail.get("attack", 0)

            elif detector_name == "nsfw":
                violation_info["nsfw_score"] = detail.get("nsfw", 0)

            elif detector_name == "bias":
                violation_info["bias_detected"] = detail.get("bias_detected", False)
                violation_info["debiased_text"] = detail.get("debiased_text", "")

            elif detector_name == "sponge_attack":
                violation_info["sponge_detected"] = detail.get("sponge_attack_detected", False)

            violations.append(violation_info)

    return violations


# ============================================================================
# ENKRYPT API FUNCTIONS
# ============================================================================

def check_with_enkrypt_api(text: str, hook_name: str = "beforeSubmitPrompt") -> tuple[bool, list, dict]:
    """
    Check text using Enkrypt AI Guardrails API.

    Args:
        text: The text to check (prompt or MCP input/output)
        hook_name: The hook name (beforeSubmitPrompt, beforeMCPExecution, afterMCPExecution, afterAgentResponse)

    Returns:
        Tuple of (should_block, violations_list, full_result)
    """
    # Check if guardrails are enabled for this hook
    if not is_hook_enabled(hook_name):
        return False, [], {"skipped": f"{hook_name} guardrails disabled"}

    # Get the block list for this hook
    block_list = get_hook_block_list(hook_name)

    try:
        import requests

        # Get policy name for this hook
        policy_name = get_hook_policy_name(hook_name)

        payload = {
            "text": text
        }

        # Debug: Log the actual request being made
        log_event("enkrypt_api_debug", {
            "url": ENKRYPT_API_URL,
            "api_key_length": len(ENKRYPT_API_KEY) if ENKRYPT_API_KEY else 0,
            "payload_text_length": len(text),
            "policy_name": policy_name,
            "config_file": str(CONFIG_FILE),
        })

        # Disable SSL warnings and verify for debugging
        import urllib3
        urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)

        resp = requests.post(
            ENKRYPT_API_URL,
            headers={
                "Content-Type": "application/json",
                "apikey": ENKRYPT_API_KEY,
                "X-Enkrypt-Policy": policy_name
            },
            json=payload,
            timeout=15,
            verify=False,  # Bypass SSL verification for debugging
        )

        # Log response status before raising
        log_event("enkrypt_api_response", {
            "status_code": resp.status_code,
            "url": resp.url,
            "response_preview": resp.text[:300] if resp.text else None,
            "request_url": ENKRYPT_API_URL,
            "headers_sent": dict(resp.request.headers) if resp.request else None,
        })

        resp.raise_for_status()
        result = resp.json()

        # Parse the response using the actual API format
        violations = parse_enkrypt_response(result, block_list)

        should_block = len(violations) > 0
        return should_block, violations, result

    except ImportError:
        log_event("enkrypt_api_error", {"error": "requests library not installed"})
        return False, [], {"error": "requests not installed"}
    except Exception as e:
        error_info = {
            "error": str(e),
            "error_type": type(e).__name__,
            "url_used": ENKRYPT_API_URL,
        }
        # Try to get response details if it's an HTTP error
        if hasattr(e, 'response') and e.response is not None:
            error_info["status_code"] = e.response.status_code
            error_info["response_text"] = e.response.text[:500] if e.response.text else None
        log_event("enkrypt_api_error", error_info)
        return False, [], {"error": str(e)}


def format_violation_message(violations: list, hook_name: str = "beforeSubmitPrompt") -> str:
    """Format a user-friendly message from violations."""
    if not violations:
        return ""

    policy_name = get_hook_policy_name(hook_name)

    messages = [f"Policy: {policy_name}\n"]

    for v in violations:
        detector = v["detector"]

        if detector == "pii":
            entities = v.get("entities", [])
            pii_found = v.get("pii_found", {})
            if pii_found:
                # Show actual PII types found
                pii_items = [f"{k}" for k in list(pii_found.keys())[:3]]
                messages.append(f"🔐 PII/Secrets detected: {', '.join(pii_items)}")
            elif entities:
                messages.append(f"🔐 PII/Secrets detected: {', '.join(str(e) for e in entities[:5])}")
            else:
                messages.append("🔐 PII/Secrets detected")

        elif detector == "injection_attack":
            attack_score = v.get("attack_score", 0)
            try:
                score_float = float(attack_score) if attack_score else 0
                if score_float:
                    messages.append(f"⚠️ Injection attack detected (confidence: {score_float:.1%})")
                else:
                    messages.append("⚠️ Injection attack pattern detected")
            except (ValueError, TypeError):
                messages.append("⚠️ Injection attack pattern detected")

        elif detector == "toxicity":
            toxicity_types = v.get("toxicity_types", [])
            score = v.get("score", "N/A")
            if toxicity_types:
                messages.append(f"🚫 Toxic content detected: {', '.join(toxicity_types)} (score: {score})")
            else:
                messages.append(f"🚫 Toxic content detected (score: {score})")

        elif detector == "nsfw":
            nsfw_score = v.get("nsfw_score", 0)
            try:
                score_float = float(nsfw_score) if nsfw_score else 0
                if score_float:
                    messages.append(f"🔞 NSFW content detected (confidence: {score_float:.1%})")
                else:
                    messages.append("🔞 NSFW content detected")
            except (ValueError, TypeError):
                messages.append("🔞 NSFW content detected")

        elif detector == "keyword_detector":
            keywords = v.get("matched_keywords", [])
            if keywords:
                messages.append(f"🚫 Banned keywords detected: {', '.join(keywords)}")
            else:
                messages.append("🚫 Banned keywords detected")

        elif detector == "policy_violation":
            violating_policy = v.get("violating_policy", "")
            explanation = v.get("explanation", "")
            if violating_policy:
                messages.append(f"📋 Policy violation: {violating_policy}")
            if explanation:
                messages.append(f"   → {explanation[:150]}")

        elif detector == "bias":
            messages.append("⚖️ Bias detected in content")

        elif detector == "sponge_attack":
            messages.append("🧽 Sponge attack detected")

        elif detector == "topic_detector":
            messages.append("📌 Off-topic or sensitive topic detected")

        else:
            messages.append(f"⚠️ {detector.replace('_', ' ').title()} detected")

    return "\n".join(messages)


# ============================================================================
# MCP TOOL CHECKING
# ============================================================================

def check_mcp_tool(tool_name: str, tool_input: str) -> tuple[str, str, str]:
    """
    Check if an MCP tool should be allowed, blocked, or require confirmation.

    Returns:
        Tuple of (permission, user_message, agent_message)
    """
    tool_name_lower = tool_name.lower()

    # Check sensitive tools from config
    for sensitive in SENSITIVE_MCP_TOOLS:
        if sensitive.lower() in tool_name_lower:
            return (
                "ask",
                f"⚠️ MCP tool '{tool_name}' requires confirmation",
                f"The MCP tool '{tool_name}' requires user approval before execution."
            )

    # Try to parse tool input and check for sensitive operations
    try:
        params = json.loads(tool_input) if tool_input else {}

        # Check for SQL operations
        if "query" in params or "sql" in params:
            query = params.get("query", "") or params.get("sql", "")
            query_upper = query.upper()

            dangerous_sql = ["DROP", "DELETE", "TRUNCATE", "UPDATE", "INSERT"]
            for keyword in dangerous_sql:
                if keyword in query_upper:
                    return (
                        "ask",
                        f"⚠️ SQL operation '{keyword}' requires confirmation",
                        f"This SQL operation modifies data and requires user approval."
                    )
    except (json.JSONDecodeError, TypeError):
        pass

    # Allow by default
    return "allow", "", ""


def analyze_mcp_result(tool_name: str, result_json: str) -> dict:
    """Analyze MCP tool result for potential issues."""
    import re

    analysis = {
        "sensitive_data_hints": [],
        "result_size": len(result_json),
        "is_error": False,
    }

    result_lower = result_json.lower()

    # Patterns to detect in MCP results
    sensitive_patterns = [
        (r"password", "password reference"),
        (r"api[_-]?key", "API key reference"),
        (r"secret", "secret reference"),
        (r"token", "token reference"),
        (r"credential", "credential reference"),
    ]

    for pattern, name in sensitive_patterns:
        if re.search(pattern, result_lower):
            analysis["sensitive_data_hints"].append(name)

    # Try to detect errors in result
    try:
        result = json.loads(result_json)
        if isinstance(result, dict):
            if result.get("error") or result.get("Error"):
                analysis["is_error"] = True
            if result.get("status") in ["error", "failed", "failure"]:
                analysis["is_error"] = True
    except (json.JSONDecodeError, TypeError):
        pass

    return analysis


# ============================================================================
# UTILITY FUNCTIONS
# ============================================================================

def reload_config():
    """Reload configuration from file (useful for dynamic updates)."""
    global CONFIG, HOOK_POLICIES, SENSITIVE_MCP_TOOLS
    CONFIG = load_config()
    HOOK_POLICIES = {
        "beforeSubmitPrompt": CONFIG.get("beforeSubmitPrompt", {}),
        "beforeMCPExecution": CONFIG.get("beforeMCPExecution", {}),
        "afterMCPExecution": CONFIG.get("afterMCPExecution", {}),
        "afterAgentResponse": CONFIG.get("afterAgentResponse", {}),
    }
    SENSITIVE_MCP_TOOLS = CONFIG.get("sensitive_mcp_tools", [])

File: hooks/vercel-ai-sdk/src/enkrypt-guardrails.ts
/**
 * Enkrypt AI Guardrails Core Module for Vercel AI SDK
 *
 * This module provides the core functionality for integrating Enkrypt AI
 * Guardrails with Vercel AI SDK applications.
 *
 * Features:
 * - HTTP client with retry logic
 * - Response parsing and violation detection
 * - Logging and metrics collection
 * - Configuration management
 */

import * as fs from 'fs';
import * as path from 'path';
import * as os from 'os';

// ============================================================================
// TYPES
// ============================================================================

export interface EnkryptApiConfig {
  url: string;
  api_key: string;
  ssl_verify: boolean;
  timeout: number;
  fail_silently: boolean;
}

export interface HookPolicy {
  enabled: boolean;
  guardrail_name: string;
  block: string[];
}

export interface GuardrailsConfig {
  enkrypt_api: EnkryptApiConfig;
  transformParams: HookPolicy;
  wrapGenerate: HookPolicy;
  wrapStream: HookPolicy;
  prepareStep: HookPolicy;
  onStepFinish: HookPolicy;
  onToolCall: HookPolicy;
  sensitive_tools: string[];
}

export interface Violation {
  detector: string;
  score: number;
  threshold: number;
  detected: boolean;
  details?: Record<string, unknown>;
}

export interface CheckResult {
  shouldBlock: boolean;
  violations: Violation[];
  rawResult: Record<string, unknown> | null;
}

export interface LogEntry {
  timestamp: string;
  hook: string;
  event: string;
  data: Record<string, unknown>;
}

export interface HookMetrics {
  totalCalls: number;
  blocked: number;
  allowed: number;
  errors: number;
  totalLatencyMs: number;
}

// ============================================================================
// CONSTANTS
// ============================================================================

const DEFAULT_API_URL = 'https://api.enkryptai.com/guardrails/policy/detect';
const DEFAULT_TIMEOUT = 15000; // 15 seconds
const LOG_DIR = path.join(os.homedir(), 'vercel-ai-sdk', 'guardrails_logs');

// Detectors that Enkrypt API can return
export const DETECTOR_TYPES = [
  'injection_attack',
  'pii',
  'toxicity',
  'nsfw',
  'bias',
  'sponge_attack',
  'keyword_detector',
  'topic_detector',
  'policy_violation',
] as const;

export type DetectorType = typeof DETECTOR_TYPES[number];

// ============================================================================
// CONFIGURATION
// ============================================================================

let config: GuardrailsConfig | null = null;

/**
 * Load configuration from file or environment
 */
export function loadConfig(configPath?: string): GuardrailsConfig {
  if (config) return config;

  const possiblePaths = [
    configPath,
    process.env.ENKRYPT_GUARDRAILS_CONFIG,
    path.join(process.cwd(), 'guardrails-config.json'),
    path.join(__dirname, '..', 'guardrails-config.json'),
  ].filter(Boolean) as string[];

  for (const p of possiblePaths) {
    try {
      if (fs.existsSync(p)) {
        const content = fs.readFileSync(p, 'utf-8');
        config = JSON.parse(content) as GuardrailsConfig;

        // Override API key from environment if set
        if (process.env.ENKRYPT_API_KEY) {
          config.enkrypt_api.api_key = process.env.ENKRYPT_API_KEY;
        }

        return config;
      }
    } catch {
      // Continue to next path
    }
  }

  // Return default config if no file found
  config = getDefaultConfig();
  return config;
}

/**
 * Get default configuration
 */
export function getDefaultConfig(): GuardrailsConfig {
  return {
    enkrypt_api: {
      url: process.env.ENKRYPT_API_URL || DEFAULT_API_URL,
      api_key: process.env.ENKRYPT_API_KEY || '',
      ssl_verify: true,
      timeout: DEFAULT_TIMEOUT,
      fail_silently: true,
    },
    transformParams: {
      enabled: true,
      guardrail_name: 'Sample Airline Guardrail',
      block: ['injection_attack', 'pii', 'toxicity', 'nsfw'],
    },
    wrapGenerate: {
      enabled: true,
      guardrail_name: 'Sample Airline Guardrail',
      block: ['pii', 'toxicity', 'nsfw'],
    },
    wrapStream: {
      enabled: true,
      guardrail_name: 'Sample Airline Guardrail',
      block: ['pii', 'toxicity', 'nsfw'],
    },
    prepareStep: {
      enabled: true,
      guardrail_name: 'Sample Airline Guardrail',
      block: ['injection_attack', 'pii'],
    },
    onStepFinish: {
      enabled: true,
      guardrail_name: 'Sample Airline Guardrail',
      block: ['pii', 'toxicity'],
    },
    onToolCall: {
      enabled: true,
      guardrail_name: 'Sample Airline Guardrail',
      block: ['injection_attack', 'pii'],
    },
    sensitive_tools: [
      'execute_sql',
      'run_command',
      'shell_*',
      'bash',
      'delete_*',
      'remove_*',
      'write_file',
      'send_email',
      'http_request',
    ],
  };
}

/**
 * Reload configuration from disk
 */
export function reloadConfig(configPath?: string): GuardrailsConfig {
  config = null;
  return loadConfig(configPath);
}

/**
 * Check if a hook is enabled
 */
export function isHookEnabled(hookName: keyof GuardrailsConfig): boolean {
  const cfg = loadConfig();
  const hookConfig = cfg[hookName];
  if (typeof hookConfig === 'object' && 'enabled' in hookConfig) {
    return hookConfig.enabled;
  }
  return false;
}

/**
 * Get block list for a hook
 */
export function getHookBlockList(hookName: keyof GuardrailsConfig): string[] {
  const cfg = loadConfig();
  const hookConfig = cfg[hookName];
  if (typeof hookConfig === 'object' && 'block' in hookConfig) {
    return hookConfig.block;
  }
  return [];
}

/**
 * Get guardrail name for a hook
 */
export function getHookGuardrailName(hookName: keyof GuardrailsConfig): string {
  const cfg = loadConfig();
  const hookConfig = cfg[hookName];
  if (typeof hookConfig === 'object' && 'guardrail_name' in hookConfig) {
    return hookConfig.guardrail_name;
  }
  return 'Default Policy';
}

// ============================================================================
// API CLIENT
// ============================================================================

/**
 * Check content with Enkrypt AI API
 */
export async function checkWithEnkryptApi(
  text: string,
  hookName: string = 'transformParams'
): Promise<CheckResult> {
  const cfg = loadConfig();
  const startTime = Date.now();

  // Skip if no API key
  if (!cfg.enkrypt_api.api_key) {
    logEvent(hookName, 'api_skipped', { reason: 'no_api_key' });
    return { shouldBlock: false, violations: [], rawResult: null };
  }

  // Skip empty text
  if (!text || !text.trim()) {
    return { shouldBlock: false, violations: [], rawResult: null };
  }

  const hookConfig = cfg[hookName as keyof GuardrailsConfig] as HookPolicy | undefined;
  const guardrailName = hookConfig?.guardrail_name || 'Default Policy';
  const blockList = hookConfig?.block || [];

  try {
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), cfg.enkrypt_api.timeout);

    const response = await fetch(cfg.enkrypt_api.url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'apikey': cfg.enkrypt_api.api_key,
        'X-Enkrypt-Policy': guardrailName,
        'X-Enkrypt-Source-Name': 'vercel-ai-sdk',
        'X-Enkrypt-Source-Event': hookName,
      },
      body: JSON.stringify({
        text: text,
      }),
      signal: controller.signal,
    });

    clearTimeout(timeoutId);

    if (!response.ok) {
      const errorText = await response.text();
      logEvent(hookName, 'api_error', {
        status: response.status,
        error: errorText,
      });

      if (cfg.enkrypt_api.fail_silently) {
        return { shouldBlock: false, violations: [], rawResult: null };
      }
      throw new Error(`Enkrypt API error: ${response.status} - ${errorText}`);
    }

    const result = await response.json();
    const latency = Date.now() - startTime;

    // Log API response
    logApiResponse(hookName, result, latency);

    // Parse violations
    const { shouldBlock, violations } = parseEnkryptResponse(result, blockList);

    // Update metrics
    updateMetrics(hookName, shouldBlock, latency);

    return { shouldBlock, violations, rawResult: result };
  } catch (error) {
    const latency = Date.now() - startTime;
    logEvent(hookName, 'api_exception', {
      error: error instanceof Error ? error.message : String(error),
      latency,
    });

    updateMetrics(hookName, false, latency, true);

    if (cfg.enkrypt_api.fail_silently) {
      return { shouldBlock: false, violations: [], rawResult: null };
    }
    throw error;
  }
}

/**
 * Parse Enkrypt API response and determine violations
 *
 * API Response Format:
 * {
 *   "summary": {
 *     "injection_attack": 1,      // 1 = detected, 0 = not detected
 *     "pii": 1,
 *     "toxicity": ["toxicity"],   // can be a list
 *     ...
 *   },
 *   "details": {
 *     "injection_attack": { "safe": "0.01", "attack": "0.99", ... },
 *     ...
 *   }
 * }
 */
export function parseEnkryptResponse(
  result: Record<string, unknown>,
  blockList: string[]
): { shouldBlock: boolean; violations: Violation[] } {
  const violations: Violation[] = [];
  let shouldBlock = false;

  const summary = result.summary as Record<string, unknown> | undefined;
  const details = result.details as Record<string, unknown> | undefined;

  if (!summary) {
    return { shouldBlock, violations };
  }

  // Map summary keys to detector names
  const detectorMapping: Record<string, string> = {
    'nsfw': 'nsfw',
    'toxicity': 'toxicity',
    'pii': 'pii',
    'injection_attack': 'injection_attack',
    'keyword_detected': 'keyword_detector',
    'policy_violation': 'policy_violation',
    'bias': 'bias',
    'sponge_attack': 'sponge_attack',
    'on_topic': 'topic_detector',
  };

  for (const [summaryKey, detectorName] of Object.entries(detectorMapping)) {
    const summaryValue = summary[summaryKey];

    // Check if this detector triggered
    let isDetected = false;
    if (typeof summaryValue === 'number' && summaryValue === 1) {
      isDetected = true;
    } else if (Array.isArray(summaryValue) && summaryValue.length > 0) {
      isDetected = true;
    } else if (typeof summaryValue === 'boolean' && summaryValue) {
      isDetected = true;
    }

    // Special case: on_topic=0 means OFF topic (violation), on_topic=1 means ON topic (ok)
    if (summaryKey === 'on_topic') {
      isDetected = typeof summaryValue === 'number' && summaryValue === 0;
    }

    if (isDetected) {
      // Get details for this detector
      const detectorDetails = (details?.[detectorName] || details?.[summaryKey] || {}) as Record<string, unknown>;

      // Extract score if available (e.g., injection_attack has "attack" score)
      let score = 0;
      if (typeof detectorDetails.attack === 'string') {
        score = parseFloat(detectorDetails.attack);
      } else if (typeof detectorDetails.score === 'number') {
        score = detectorDetails.score;
      }

      const violation: Violation = {
        detector: detectorName,
        score,
        threshold: 0.5,
        detected: true,
        details: detectorDetails,
      };
      violations.push(violation);

      // Check if this detector should trigger blocking
      if (blockList.includes(detectorName)) {
        shouldBlock = true;
      }
    }
  }

  return { shouldBlock, violations };
}

/**
 * Format violation message for user display
 */
export function formatViolationMessage(violations: Violation[], hookName: string = 'guardrails'): string {
  if (violations.length === 0) {
    return '';
  }

  const detectorNames = violations.map(v => v.detector).join(', ');
  return `[Enkrypt Guardrails - ${hookName}] Content blocked due to: ${detectorNames}`;
}

// ============================================================================
// TOOL UTILITIES
// ============================================================================

/**
 * Check if a tool is considered sensitive
 */
export function isSensitiveTool(toolName: string): boolean {
  const cfg = loadConfig();
  const patterns = cfg.sensitive_tools || [];

  for (const pattern of patterns) {
    if (pattern.includes('*')) {
      // Wildcard pattern
      const regex = new RegExp('^' + pattern.replace(/\*/g, '.*') + '$');
      if (regex.test(toolName)) {
        return true;
      }
    } else {
      // Exact match or prefix match
      if (toolName === pattern || toolName.startsWith(pattern)) {
        return true;
      }
    }
  }

  return false;
}

// ============================================================================
// LOGGING
// ============================================================================

let logBuffer: LogEntry[] = [];
const LOG_FLUSH_INTERVAL = 5000; // 5 seconds
const LOG_BUFFER_SIZE = 100;

// Ensure log directory exists
function ensureLogDir(): void {
  try {
    if (!fs.existsSync(LOG_DIR)) {
      fs.mkdirSync(LOG_DIR, { recursive: true });
    }
  } catch {
    // Ignore errors - logging is optional
  }
}

/**
 * Log an event
 */
export function logEvent(
  hook: string,
  event: string,
  data: Record<string, unknown> = {}
): void {
  const entry: LogEntry = {
    timestamp: new Date().toISOString(),
    hook,
    event,
    data,
  };

  logBuffer.push(entry);

  if (logBuffer.length >= LOG_BUFFER_SIZE) {
    flushLogs();
  }
}

/**
 * Log API response details
 */
function logApiResponse(
  hook: string,
  result: Record<string, unknown>,
  latency: number
): void {
  ensureLogDir();

  const entry = {
    timestamp: new Date().toISOString(),
    hook,
    latency_ms: latency,
    result_preview: JSON.stringify(result).slice(0, 500),
  };

  try {
    const logPath = path.join(LOG_DIR, 'enkrypt_api_response.jsonl');
    fs.appendFileSync(logPath, JSON.stringify(entry) + '\n');
  } catch {
    // Ignore logging errors
  }
}

/**
 * Log security alert
 */
export function logSecurityAlert(
  alertType: string,
  details: Record<string, unknown>,
  context: Record<string, unknown> = {}
): void {
  ensureLogDir();

  const entry = {
    timestamp: new Date().toISOString(),
    alert_type: alertType,
    details,
    context,
  };

  try {
    const logPath = path.join(LOG_DIR, 'security_alerts.jsonl');
    fs.appendFileSync(logPath, JSON.stringify(entry) + '\n');
  } catch {
    // Ignore logging errors
  }
}

/**
 * Flush log buffer to disk
 */
export function flushLogs(): void {
  if (logBuffer.length === 0) return;

  ensureLogDir();

  try {
    const logPath = path.join(LOG_DIR, 'combined_audit.jsonl');
    const content = logBuffer.map(e => JSON.stringify(e)).join('\n') + '\n';
    fs.appendFileSync(logPath, content);
    logBuffer = [];
  } catch {
    // Clear buffer even on error to prevent memory issues
    logBuffer = [];
  }
}

// Flush logs periodically
if (typeof setInterval !== 'undefined') {
  setInterval(flushLogs, LOG_FLUSH_INTERVAL);
}

// ============================================================================
// METRICS
// ============================================================================

const metrics: Map<string, HookMetrics> = new Map();

/**
 * Update metrics for a hook
 */
function updateMetrics(
  hookName: string,
  blocked: boolean,
  latencyMs: number,
  isError: boolean = false
): void {
  let hookMetrics = metrics.get(hookName);

  if (!hookMetrics) {
    hookMetrics = {
      totalCalls: 0,
      blocked: 0,
      allowed: 0,
      errors: 0,
      totalLatencyMs: 0,
    };
    metrics.set(hookName, hookMetrics);
  }

  hookMetrics.totalCalls++;
  hookMetrics.totalLatencyMs += latencyMs;

  if (isError) {
    hookMetrics.errors++;
  } else if (blocked) {
    hookMetrics.blocked++;
  } else {
    hookMetrics.allowed++;
  }
}

/**
 * Get metrics for all hooks
 */
export function getMetrics(): Record<string, HookMetrics & { avgLatencyMs: number }> {
  const result: Record<string, HookMetrics & { avgLatencyMs: number }> = {};

  metrics.forEach((m, hookName) => {
    result[hookName] = {
      ...m,
      avgLatencyMs: m.totalCalls > 0 ? m.totalLatencyMs / m.totalCalls : 0,
    };
  });

  return result;
}

/**
 * Reset metrics
 */
export function resetMetrics(): void {
  metrics.clear();
}

// ============================================================================
// EXCEPTIONS
// ============================================================================

/**
 * Error thrown when guardrails block a request
 */
export class GuardrailsViolationError extends Error {
  violations: Violation[];
  hookName: string;

  constructor(message: string, violations: Violation[], hookName: string) {
    super(message);
    this.name = 'GuardrailsViolationError';
    this.violations = violations;
    this.hookName = hookName;
  }
}
